[14.0][IMP] stock_mts_mto_rule: add split option

TDu opened the pull request:
With this change a new option is available on the stock rule. It allows to use the MTS rule only if the full quantity is avaiable.

mt-software-de commented on stock_mts_mto_rule/models/stock_rule.py:
```diff
@@ -26,14 +35,18 @@ def _check_mts_mto_rule(self):
         src_location_id = self.mts_rule_id.location_src_id.id
         product_location = product.with_context(location=src_location_id)
-        virtual_available = product_location.virtual_available
+        virtual_available = self._get_available_quantity_mts_mto_rule(product_location)
```
Could we simplify this computation?
And adding the check for the mts_mto_split_quantity_rule here instead of in _run_split_procurement?

```suggestion
        virtual_available = self._get_available_quantity_mts_mto_rule(product_location)
        qty_available = product.uom_id._compute_quantity(virtual_available, product_uom)
        qty_mto = product_qty - qty_available
        if float_compare(qty_mto, 0, precision_digits=precision) <= 0:
            return 0
        if rule.mts_mto_split_quantity_rule == "full":
            return product_qty
        return qty_mto
```

TDu replied:
I am not sure of this.
We don't have access to `rule` in this function, and I don't like changing the signature of an existing function.
And it also makes the diff bigger.

mt-software-de replied:
we have access to the rule it is self.

mt-software-de replied:
yes it makes the diff bigger, but i think it makes it much more understandable.
The method _run_split_procurement would not change. 

Because we only need to change the returned qty.

TDu replied:
> we have access to the rule it is self.

oh :blush:  you are right

mt-software-de replied:
@TDu do you want to simplify the computation? :-)

mt-software-de replied:
ping @TDu ?

TDu replied:
Thanks for your review. But I would like to keep the computation as it is, now.

mt-software-de commented on stock_mts_mto_rule/models/stock_rule.py:
```diff
@@ -15,6 +15,15 @@ class StockRule(models.Model):
     )
     mts_rule_id = fields.Many2one("stock.rule", string="MTS Rule", check_company=True)
     mto_rule_id = fields.Many2one("stock.rule", string="MTO Rule", check_company=True)
+    mts_mto_split_quantity_rule = fields.Selection(
+        string="MTS/MTO split rule",
+        selection=[("diff", "Difference"), ("full", "No Split")],
+        default="diff",
+        help=(
+            "With `Difference` the MTS rule will be used for any available quantity.\n"
+            "With`No Split` only if the full quantity is avaiable."
+        ),
+    )
```
```suggestion
    mts_quantity_rule = fields.Selection(
        string="MTS quantity rule",
        selection=[("available", "Available"), ("full", "No Split")],
        default="available",
        help=(
            "With `Available` the MTS rule will be used for any available quantity.\n"
            "With`No Split` the MTS rule is only used, if the full requested quantity is available"
        ),
    )
```

TDu replied:
Updated

florian-dacosta commented:
@TDu Isn't it the goal of the native procure method `mts_else_mto` ?
I did not check this recently, but I had in mind the stock_mts_mto_rule module was only still usefull for the case we want to split between mts and mto when there is an available qty in stock but is not enough for the whole procurement. Else, native Odoo system do the job. Maybe I am missing something though

TDu commented:
> @TDu Isn't it the goal of the native procure method `mts_else_mto` ? I did not check this recently, but I had in mind the stock_mts_mto_rule module was only still usefull for the case we want to split between mts and mto when there is an available qty in stock but is not enough for the whole procurement. Else, native Odoo system do the job. Maybe I am missing something though

I don't think I can fix my need with the `mts_else_mto` method. Because I can not choose the next procurement method, in my case if the stock is not available I want to call the dropshipping rule.

mt-software-de reviewed (approved):
LGTM

jbaudoux commented:
@TDu Please squash commits

TDu commented:
Squashed

jbaudoux commented:
@TDu I forgot to ask you to also drop changes to generated files (README.rst & index.html)to ease fwd port

TDu committed:
stock_mts_mto_rule: add split option

With this change a new option is available on the stock rule.
It allows to use the MTS rule only if the full quantity is avaiable.

jbaudoux commented:
/ocabot merge minor
